types/js-cookie/index.d.ts

```diff
@@ -32,6 +33,18 @@ declare namespace Cookies {
          * secure protocol (https). Defaults to false.
          */
         secure?: boolean;
+
+        /**
+         * Define the first party cookie policy.
+         * @see: https://tools.ietf.org/html/draft-west-first-party-cookies-07
+         */
+        samesite?: 'Strict' | 'Lax';
+
+        /**
+         * An attribute which will be serialized, conformably to RFC 6265
+         * section 5.2.
+         */
+        [property: string]: any;
```

Comment:
I'm not sure it's a good idea to add this. What is the use case? Wouldn't it prevent catching mistakes when one of the supported attributes is mistyped?

Reply:
I don't think it's the role of types definition to validate or constraint values. Sure adding known standardized attributes is a nice added value, but you also have to accuratly represent the code ability.

Right now the library is compatible with `SameSite` and also with other attributes who may be added in the future. The only thing breaking it is too restrictive type definition.

If in a few years navigators decide to add let's say `SpdyOnly` or whatever, the library will be fine, types definition should also.

Reply:
Restricting the properties is an API design decision that should be made in the library. Types definition, IMHO, should be about describing and providing type safety, not adding our own opinions on API design.

Reply:
Normally I'd agree about it being a library decision, but in this case it's more an implementation detail of the library, and the API design decision is ultimately made by the HTTP Cookie specification implemented by browsers. If some other attribute is hypothetically added in the future, the type definitions should be updated to correctly suggest and constrain this property (just as is being done now for SameSite).

Reply:
I'm sorry to insist, but it's not an implementation detail, it's by design.

1. The RFC has the grammar for it (https://tools.ietf.org/html/rfc6265#section-4.1.1):
```ABNF
 cookie-av         = expires-av / max-age-av / domain-av /
                     path-av / secure-av / httponly-av /
                     extension-av
 expires-av        = "Expires=" sane-cookie-date
 max-age-av        = "Max-Age=" non-zero-digit *DIGIT
 domain-av         = "Domain=" domain-value
 path-av           = "Path=" path-value
 secure-av         = "Secure"
 httponly-av       = "HttpOnly"
 extension-av      = <any CHAR except CTLs or ";">
```
2. There's an issue for it (https://github.com/js-cookie/js-cookie/issues/276)
3. And also a merged PR complete with explicit unit test for unofficial attributes (https://github.com/js-cookie/js-cookie/pull/293)

Reply:
Okay, I guess it makes sense then. Shouldn't the value at least have a type of `string`, though, instead of `any`?

Reply:
Well you can pass:
* `string` which would be used as is
* `number` which would be casted to string
* boolean `true`, the attribute will be printed but won't have value (the loop will hit the second `continue`)

Other values can technically also be passed but don't really make sense:
* boolean `false`, will skip the attribute (the loop will hit the first `continue`)
* `array` and `function` will be casted to string

Kinda grey area:
* `object` could potentially make sense if the object implements `toString`

Reply:
If the property can only hold a string value, I think it should be declared as such. Devs can call `toString` themselves on objects/numbers if they want to store them. Otherwise it's confusing if you get a different value out than you put in.

Reply:
So `string | number | boolean` ?

Reply:
No, only `string` (just like the value parameter on the `set` method has a string type, even though technically in JavaScript you could pass other things to it which would be cast to a string).

Reply:
Accepting only `string` won't do the trick because you need at least `string | boolean` to declare value-less attributes.

Also I just tested it and the generic property declaration have to at least be compatible with all other attributes declaration or else other attributes will throw `ts(2411)` errors.
So the bare minimum is `[property: string]: string | number | boolean | Date | undefined;`... might as well keep `any`

https://www.typescriptlang.org/docs/handbook/interfaces.html#indexable-types

Reply:
Thank you for your time @theodorejb